lib/functions.php

```diff
@@ -350,3 +352,45 @@ function put(string $path, string $contents): Promise
 {
     return filesystem()->put($path, $contents);
 }
+
+/**
+ * Asynchronously lock a file
+ * Resolves with a callbable that MUST eventually be called in order to release the lock.
+ *
+ * @param string  $file      File to lock
+ * @param integer $operation Locking mode, one of \LOCK_SH or \LOCK_EX (see PHP flock docs)
+ * @param integer $polling   Polling interval for lock in milliseconds
+ *
+ * @return \Amp\Promise Resolves with a callbable that MUST eventually be called in order to release the lock.
+ */
+function lock(string $file, int $operation, int $polling = 100): Promise
```

Comment:
`$operation` is basically a boolean operator here, I guess it makes sense to turn this into two methods: `lockShared(string, int)` and `lockExclusively(string, int)`